source/ada/lsp-predefined_completion.adb

```diff
@@ -0,0 +1,214 @@
+
+   procedure Load_Predefined_Items
+     (Trace   : Trace_Handle;
+      Version : Ada_Version_Type)
+   is
+      Root : JSON_Value := JSON_Null;
+
+      procedure Load_Items_From_Key
+        (Key   : String;
+         Items : in out CompletionItem_Vector);
+
+      procedure Load_Item
+        (Value : GNATCOLL.JSON.JSON_Value;
+         Items : in out CompletionItem_Vector);
+
+      ---------------
+      -- Load_Item --
+      ---------------
+
+      procedure Load_Item
+        (Value : GNATCOLL.JSON.JSON_Value;
+         Items : in out CompletionItem_Vector)
+      is
+         Item : CompletionItem;
+      begin
+         Item.insertTextFormat := (True, PlainText);
+         Item.label := To_LSP_String (String'(Value.Get ("_name")));
+         Item.detail := (True, To_LSP_String (String'(Value.Get ("_origin"))));
+         Item.insertText := (True, Item.label);
+         Item.documentation :=
+           (Is_Set => True,
+            Value  => String_Or_MarkupContent'
+              (Is_String => True,
+               String    => To_LSP_String
+                 (String'(Value.Get ("DOC")))));
+         Item.kind := (True, Keyword);
+
+         Items.Append (Item);
+      end Load_Item;
+
+      -------------------------
+      -- Load_Items_From_Key --
+      -------------------------
+
+      procedure Load_Items_From_Key
+        (Key   : String;
+         Items : in out CompletionItem_Vector)
+      is
+         Array_Node : constant JSON_Array := Root.Get (Key);
+      begin
+         for Value of Array_Node loop
+            Load_Item (Value, Items);
+         end loop;
+      end Load_Items_From_Key;
+
+   begin
+      --  TODO: load the proper database once we have one for each Ada version
+      case Version is
+         when Ada_83 =>
+            Root := Read (LSP.Predefined_Completion.Ada2012.Db);
+         when Ada_95 =>
+            Root := Read (LSP.Predefined_Completion.Ada2012.Db);
+         when Ada_2005 =>
+            Root := Read (LSP.Predefined_Completion.Ada2012.Db);
+         when Ada_2012 =>
+            Root := Read (LSP.Predefined_Completion.Ada2012.Db);
+         when Ada_2020 =>
+            Root := Read (LSP.Predefined_Completion.Ada2012.Db);
+      end case;
+
+      Root := Root.Get ("PREDEFINED_ADA");
+
+      if Root.Has_Field ("ASPECT") then
+         Load_Items_From_Key (Key => "ASPECT", Items => Aspects);
+      end if;
+
+      if Root.Has_Field ("ATTRIBUTE") then
+         Load_Items_From_Key (Key => "ATTRIBUTE", Items => Attributes);
+      end if;
+
+      if Root.Has_Field ("PRAGMA") then
+         Load_Items_From_Key (Key => "PRAGMA", Items => Pragmas);
+      end if;
+
+   exception
+      when E : others =>
+         Trace.Trace (E);
+         return;
+   end Load_Predefined_Items;
+
+   -----------------------------------
+   -- Load_Predefined_Completion_Db --
+   -----------------------------------
+
+   procedure Load_Predefined_Completion_Db (Trace : Trace_Handle) is
```

Comment:
Why do we need to launch the `gnatmake` on each DidChangeConfiguration? I don't see how its output could change

Reply:
It's not a bad idea actually IMP, since the user may have changed its PATH before reloading his project for instance, no?

Reply:
AFAIK, there is no way to change environment variables in another process after its start, so GnatStutio can't change PATH in running ALS. The only way is to restart ALS. I guess, ALS doesn't change its environment variables during the execution. Call to `setenv` isn't thread-safe on Linux and should be avoided in ALS.

Reply:
Ok, I'll call this only at startup then.